notebooks/Tutorial_Top-K_Multidimensional_Motif_and_Matches_Discovery.ipynb

```diff
@@ -0,0 +1,250 @@
+{
```

Comment:
<p>What if instead of setting things to <code>np.nan</code>, set things to random noise?</p><p> </p><pre class="ql-syntax" spellcheck="false">m = 4 * 60

padded_stddev = np.zeros(len(df))
for colname in df.columns:
&nbsp; &nbsp; noise = np.random.uniform(0.0, 0.1, size=len(df))
&nbsp; &nbsp; stddev = core.rolling_nanstd(df.loc[:, colname].values, m)
&nbsp; &nbsp; padded_stddev[-len(stddev):] = stddev
&nbsp; &nbsp; modified_values = df.loc[:, colname].where(padded_stddev != 0.0, noise)
&nbsp; &nbsp; df = df.assign(**{colname: modified_values})
&nbsp; &nbsp; # # Also eliminate 0 regions
&nbsp; &nbsp; zero_idx = df.loc[:, colname].where(df.loc[:, colname] != 0, noise)
&nbsp; &nbsp; df = df.assign(**{colname: zero_idx})  
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/557/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Reply:
After this, I did:

```
mps, indices = stumpy.mstump(df.astype(float), m)
motifs_distances, motifs_indices, motifs_subspaces, motifs_mdls = mmotifs(
    df.astype(float), mps, indices, max_distance=np.inf, max_matches=2
)
``` 

I think we need to allow `max_motifs=5` if we wanted to find more motifs as the default is `max_motifs=1`

Reply:
>What if instead of setting things to np.nan, set things to random noise

That's a pretty good idea! Thank you!

>I think we need to allow max_motifs=5 if we wanted to find more motifs as the default is max_motifs=1 

Yes, I agree. But since we want to show as much of the `mmotifs` function as we can, I would do everything step by step. Or what would you say? 

My suggestion would be to proceed as follows:
I would try to set different parameter combinations in `mmotifs` and try to find different things. For example we could try to find the first motif unconstrained (as it is now), explain how to choose the parameters in `mmotifs` to do so, plot the found motif and explain what we have done, in the first step. Afterwards we could do exactly the same steps but try to find a 2-dimensional motif (instead of the unconstrained one) for example. If we also have the 2-dimensional motif, we could search for its nearest neighbors (matches) and so on. And after that I would try to find some kind of story that explains what we are doing in each step and what we are finding / trying to find.  

That means, that I would try to call `mmotifs` in different ways in order to explain our input parameters. Then I would plot what we have found and explain our findings. We also could show the MDL results after the unconstrained search, to show the user what `mmotifs` is doing if no `k` is set and explain that. And so on ...
I think it might be easiest to do the programming steps first, look what parameter combinations lead to meaningful results and write the context/explanation for the jupyter notebook at the end since it depends on the parameter settings. 

What do you think?

```
num_motifs = motifs_distances.shape[0]

for motif_num in range(num_motifs):
    motif_indices = motifs_indices[motif_num]
    mdl = motifs_mdls[motif_num]
    S = motifs_subspaces[motif_num]
    subspace = [df.columns.values[s] for s in S]

    k = len(S)
    f, axs = plt.subplots(k, 1, figsize=(20, 20), sharex=True)
    plt.suptitle(f"The {k}-dimensional motif pair")

    for j, s in enumerate(S):
        time_series = df[df.columns[s]]
        if k > 1:
            ax = axs[j]
        else:
            ax = axs
        ax.plot(time_series, c="0.75")
        ax.set_title(df.columns.values[s])

        motif_idx = motif_indices[0]
        nn_idx = motif_indices[1]
        ax.plot(time_series[motif_idx : motif_idx + m], c="r", linewidth=4)
        ax.plot(time_series[nn_idx : nn_idx + m], c="r", linewidth=4)
        # ax.axvline(x=motif_idx, linestyle='dashed', c='k', linewidth=2)
        # ax.axvline(x=nn_idx, linestyle='dashed', c='k', linewidth=2)

        # Only relevant if you want to find further matches
        # Set motif_indices[2:] to avoid double counting the motif pair
        [ax.plot(time_series[match : match + m]) for match in motif_indices[2:]]
```

I think it might be useful to wrap that code in a separate function for this tutorial since it will be repeated many times. I'm asking because I haven't seen any function definitions in your tutorials yet. I would define the function at the beginning of the tutorial and explain that it will be used many times to visualize our results/findings.

Reply:
I think the first demonstration should always be calling the function with default parameters as it is the easiest and most straightforward:

```
mps, indices = stumpy.mstump(df.astype(float), m)
motifs_distances, motifs_indices, motifs_subspaces, motifs_mdls = mmotifs(df.astype(float), mps, indices)
```

So, you get something reasonable back with very little effort. Then, I would motivate the next step with "but this only gave us one motif and, at most, 10 matches", so how would we find more motifs beyond this first "top" one? And, by asking logical questions, we can motivate the introduction/need for the other parameters.

>  I'm asking because I haven't seen any function definitions in your tutorials yet. I would define the function at the beginning of the tutorial and explain that it will be used many times to visualize our results/findings.

You are very observant and thank you for asking! In general, for STUMPY code, I completely agree that we should write reusable functions rather than repeating ourselves everywhere. However, for tutorials, I personally hate needing to scroll all the way up to the top of the notebook in order to remember what the function does. I just can't keep it all in my head and I'd rather having all of the relevant code all in one place for a tutorial. Let me be absolutely clear that I'm not claiming to be "right" here but I think that by not defining functions in tutorials makes the barrier to entry a little lower for those who are new to Python. Considering that you've been a consumer of the tutorials, I'm open to your feedback and criticism as well! How do you feel about it? What are your thoughts?

Reply:
>You are very observant and thank you for asking! In general, for STUMPY code, I completely agree that we should write reusable functions rather than repeating ourselves everywhere. However, for tutorials, I personally hate needing to scroll all the way up to the top of the notebook in order to remember what the function does. I just can't keep it all in my head and I'd rather having all of the relevant code all in one place for a tutorial. Let me be absolutely clear that I'm not claiming to be "right" here but I think that by not defining functions in tutorials makes the barrier to entry a little lower for those who are new to Python. So the repeated code in our tutorials was not a lazy choice but one that was aiming to provide a more "inclusive"/"locally self-contained" programming approach. Considering that you've been a consumer of the tutorials, I'm open to your feedback and criticism as well! How do you feel about it? What are your thoughts?

Since different visualizations were always chosen in the previous tutorials, I thought this made sense. However, I think that in a tutorial where you want to explain the functionality of a function by setting different parameters it may be clearer to define one visualization function since we are using exactly the same code many times (which was not the case in the previous tutorials) in order to explain how everything works and what the differences between the function calls are. 

>However, for tutorials, I personally hate needing to scroll all the way up to the top of the notebook in order to remember what the function does. I just can't keep it all in my head and I'd rather having all of the relevant code all in one place for a tutorial.

Generally I agree with you here. But in this specific case I think it might be easier/clearer to define a `visualize()` or `show_matches()` function. Such a function is so simple that I think the name says enough about it that everyone knows what is happening without having to scroll up for it 🤔 

>I think the first demonstration should always be calling the function with default parameters as it is the easiest and most straightforward:

Yes, that sounds reasonable! I wanted to make sure at the beginning that I actually get exactly one motif pair. I'll take a look if we get a solution with the default parameters (in my testing data I had to at least set the `max_distance` parameter in order to find a motif).

Reply:
> However, I think that in a tutorial where you want to explain the functionality of a function by setting different parameters it may be clearer to define one visualization function since we are using exactly the same code many times (which was not the case in the previous tutorials) in order to explain how everything works and what the differences between the function calls are.
> Such a function is so simple that I think the name says enough about it that everyone knows what is happening without having to scroll up for it

@SaVoAMP I trust your judgement. Go for it and let's see how it turns out!

> Yes, that sounds reasonable! I wanted to make sure at the beginning that I actually get exactly one motif pair. I'll take a look if we get a solution with the default parameters (in my testing data I had to at least set the max_distance parameter in order to find a motif).

Sounds good!

Reply:
I'm not sure why but every time I rerun the notebook, I get different outputs, for example:
![image](https://user-images.githubusercontent.com/95933072/157873252-85ce3777-1577-48aa-a88e-6b83e2d06a86.png)
![image](https://user-images.githubusercontent.com/95933072/157873292-30f5c18d-f401-4f6c-a298-4b38a5240045.png)

Reply:
<p>This does sound strange. It implies that you may be overwriting your original data structures. Are your <code>otifs_distances, motifs_indices, motifs_subspaces, motifs_mdls</code> still the same when you re-run the notebook or are they changing after each re-run? </p>